crates/fuel-core/src/service/adapters/graphql_api/off_chain.rs

```diff
@@ -282,6 +304,220 @@ impl OffChainDatabase for OffChainIterableKeyValueView {
                 .into_boxed()
         }
     }
+
+    fn coins_to_spend(
+        &self,
+        owner: &Address,
+        asset_id: &AssetId,
+        target_amount: u64,
+        max_coins: u16,
+        excluded_ids: &ExcludedKeysAsBytes,
+    ) -> StorageResult<Vec<CoinId>> {
+        let prefix: Vec<_> = NON_RETRYABLE_BYTE
+            .as_ref()
+            .iter()
+            .copied()
+            .chain(owner.as_ref().iter().copied())
+            .chain(asset_id.as_ref().iter().copied())
+            .collect();
+
+        let big_first_iter = self.iter_all_filtered::<CoinsToSpendIndex, _>(
+            Some(&prefix),
+            None,
+            Some(IterDirection::Reverse),
+        );
+
+        let dust_first_iter = self.iter_all_filtered::<CoinsToSpendIndex, _>(
+            Some(&prefix),
+            None,
+            Some(IterDirection::Forward),
+        );
+
+        let selected_iter = select_coins_to_spend(
+            big_first_iter,
+            dust_first_iter,
+            target_amount,
+            max_coins,
+            excluded_ids,
+        )?;
+
+        into_coin_id(selected_iter, max_coins as usize)
```

Comment:
Do we want have this logic here on adapter level or in the `graph_ql` module? I think the database just can expose `self.iter_all_filtered::<CoinsToSpendIndex` functionality and the algorithm itself will leave in the `graph_ql` module, allowing us keep the business logic in the module instead of putting it to the `fuel-core`

Reply:
Is this more or less what you had in mind? Commit: 77f8caf0d6e7839f73097ba09a215aedad0da09f

I decided to not expose `iter_all_filtered::<CoinsToSpendIndex, _>` directly, but keep some guardrails around prefix structure and iterator directions so that the caller doesn't have to worry about it.

Reply:
Please note that I also updated location of the code in this commit: https://github.com/FuelLabs/fuel-core/pull/2463/commits/7459c3ce744496ca84263e2b01e9f4eacd33604f